docs/designs/dynamic_sharding_for_images.md

```diff
@@ -0,0 +1,200 @@
+    "{'image_id': 203564,
+    'id': 37,
+    'caption': 'A bicycle replica with a clock as the front wheel.'}"
+    ```
+
+    In elastic training, we need to split the training data into shards and
+    assign those shards to workers. When the worker fails, we need to
+    reassign uncompleted shards of the worker to other alive workers.
+
+## Shard Definition
+
+In ElasticDL, we define the shard information `start`, `end` and `shard_name`
+in the task. We can define the shard independently and and expose the
+shard for users to create their datasets.
+
+```proto
+message shard {
+  // The storage name for data, such as a MaxCompute Table,
+  // CSV file with image paths.
+  string name = 1;
+
+  // Starting and ending (non-inclusive) record number.
+  int64 start = 2;
+  int64 end = 3;
+}
+```
+
+In order to split the training data into shards, we must get the size of the
+training data firstly. For the former 2 ways, we can we can get the size by
+`len(os.listdir("images"))` and reading the CSV file. For simplicity, we also
+can store the image names in a CSV file for the first way to store images.
+So, the name of the shard is the CSV file. The start and end indices are
+the line number of the CSV file. When the worker gets the shard, it can read
+the images by the lines in the CSV file.
+
+It is difficult to get the size because we don't know the format of the
+description. So, users need to indicate the size of training data. We can
+design a function in the model definition and users implement the function
+to return the size.
+
+```python
+def get_training_size():
+    with open("annotations/captions_val2014.json") as f:
+        data = json.load(f)
+    return len(data["images"])
+```
+
+Then, the master will call the function to get the size and split the
+training data into shards by the size. The shard message will only contains
+the start and end index. Users need to read the image information according
+to the index by themselves.
+
+## The Worker Creates the Dataset using Shards
+
+### APIs to Fetch Shards
+
+```python
+class DynamicShardingManager(object):
+    def __init__(self):
+        master_addr = os.getenv("MASTER_ADDR")
+        worker_id = os.getenv("WORKER_ID")
+        self.master_client = MasterClient(
+            build_channel(master_addr), worker_id
+        )
+        self._pending_tasks = []
+        self.record_count = 0
+
+    def fetch_shard(self):
+        shard = self.master_client.get_task().shard
+        self.record_count += shard.end - shard.start
+        self._pending_shards.append(shard)
+        return shard
+
+    def report_shard_done(self):
+        task = self._pending_tasks.pop()
+        self.master_client.report_task_result(task.id)
+```
+
+### Create Dataset Using TensorFlow
+
+```python
+import tensorflow as tf
+
+global dynamic_sharding = DynamicShardingManager()
+
+class DynamicShardingHook(tf.train.SessionRunHook):
+    def __init__(self, num_worker, num_shards=100, max_steps=None):
```

Comment:
what is `num_shards` ? And why num_worker and max_step are needed?

Reply:
Those two arguments are redundant and I will delete them.